RA8x2 ethernet driver support

duynguyenxa opened the pull request:
This PR introduce the support of Ethernet driver on RA8x2 devices base on eswm, rmac and rmac phy HWIP
This PR also add a new phy driver support for GPY111 chip

Commit:
drivers: entropy: add trng for ra8x2

This commit add entropy support for ra8x2.

Signed-off-by: Nhat Ta <nhat-minh.ta.yn@bp.renesas.com>

Commit:
drivers: ethernet: phy: add Maxlinear GPY111 phy driver

This commit add phy driver support for Maxlinear GPY111. This driver
implements vendor specific behaviour rgmii delay time setup, which is
not present in the generic MII driver.

Signed-off-by: Nhat Ta <nhat-minh.ta.yn@bp.renesas.com>

Commit:
drivers: mdio: initial support for mdio rmac driver

This commit add mdio driver support for ra8x2.

Signed-off-by: Nhat Ta <nhat-minh.ta.yn@bp.renesas.com>

Commit:
drivers: ethernet: initial support for ethernet rmac driver

This commit add ethernet driver support for ra8x2.

Signed-off-by: Nhat Ta <nhat-minh.ta.yn@bp.renesas.com>

Commit:
drivers: ethernet: Put DMA buffer to nocache

This commit put ethernet buffers and descriptors to nocache .

Signed-off-by: Nhat Ta <nhat-minh.ta.yn@bp.renesas.com>

Commit:
dts: arm: renesas: ra: ra8: add Ethernet for ra8x2

This commit define eswm module for ra8x2 that includes ethernet
and mdio nodes.

Signed-off-by: Nhat Ta <nhat-minh.ta.yn@bp.renesas.com>

Commit:
manifest: update hal_renesas revision

Change hal_renesas revision to support ethernet for ra8x2

Signed-off-by: Nhat Ta <nhat-minh.ta.yn@bp.renesas.com>

Commit:
board: renesas: add ethernet for ra8x2 boards

This commit add ethernet support for ek_ra8p1, ek_ra8m2,
ek_ra8d2, mck_ra8t2.

Signed-off-by: Nhat Ta <nhat-minh.ta.yn@bp.renesas.com>

jukkar commented on drivers/ethernet/eth_renesas_ra_rmac.c:
```diff
@@ -0,0 +1,733 @@
+#if defined(CONFIG_ETH_RENESAS_RA_USE_HW_WRITEBACK)
+	struct k_sem tx_sem;
+#endif
+	uint8_t phy_link_up;
```
This looks to be used as a boolean in the code, so why to have `bool` type?

jukkar reviewed (commented):
Is the driver compile tested by the CI? If not, then I suggest you add the tests so that the code does not bit rot easily.

maass-hamburg commented on drivers/ethernet/eth_renesas_ra_rmac.c:
```diff
@@ -0,0 +1,745 @@
+	}
+
+rx_end:
+	if (ret) {
```
```suggestion
	if (ret < 0) {
```
no implicit conversion to bool

duynguyenxa replied:
@maass-hamburg , I've fixed the code as your suggestion

maass-hamburg commented on drivers/ethernet/phy/phy_maxlinear_gpy111.c:
```diff
@@ -0,0 +1,631 @@
+#define RESET_GPIO(n)                                                                              \
+	.reset_gpio = GPIO_DT_SPEC_INST_GET_OR(n, reset_gpios, {0}),                               \
+	.reset_assert_duration_us = DT_INST_PROP_OR(n, reset_assert_duration_us, 0),               \
+	.reset_deassertion_timeout_ms = DT_INST_PROP_OR(n, reset_deassertion_timeout_ms, 0),
```
these were dt props in the phy_mii driver, because these are different on different models, but for this driver you can look them up and make them defines at the top of the file and directly use them, no need then to save them in the config struct.

maass-hamburg commented on drivers/ethernet/phy/phy_maxlinear_gpy111.c:
```diff
@@ -0,0 +1,631 @@
+		     "At least one valid speed must be configured for this driver");               \
+                                                                                                   \
+	static const struct maxlinear_gpy111_dev_config maxlinear_gpy111_dev_config_##n = {        \
+		.phy_addr = 0,                                                                     \
```
this should come from the `reg` prop of the dt

duynguyenxa replied:
I changed this to get dt

duynguyenxa committed:
soc: renesas: ra: Select ARM MPU as default for RA8x2

Make ARM MPU configuration default for RA8x2 family to support
CPU caching

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

maass-hamburg commented:
@duynguyenxa please address all the comments and requested changes, before re-requesting reviews.

duynguyenxa commented on drivers/ethernet/phy/phy_maxlinear_gpy111.c:
```diff
@@ -0,0 +1,631 @@
+			return ret;
+		}
+	}
+#endif /* ANY_INT_GPIO */
```
It's removed

duynguyenxa commented on drivers/ethernet/mdio/mdio_renesas_ra_rmac.c:
```diff
@@ -0,0 +1,195 @@
+	return ret;
+}
+
+static int mdio_renesas_ra_write(const struct device *dev, uint8_t prtad, uint8_t regad,
```
Yes, I understand. I have already made several modifications to the HAL driver so that all PHY communication is handled exclusively via Zephyr’s Ethernet PHY API. However, we also want to keep the HAL changes as minimal as possible to reduce maintenance overhead when updating to newer HAL versions, and this approach satisfies that requirement.

maass-hamburg replied:
Simply use 
```
#ifndef __ZEPHYR__
  // non zephyr code
#endif /* __ZEPHYR__ */
```
in the hal code to "remove" the unneeded parts, this shouldn't make it too difficult, when there are updates of the HAL. You could even upstream these parts internally.

But in general using a HAL, especially for ethernet, often leads to sub-optimal performance, as data is often copied more often then necessary. At least if the hal don't uses zephyrs `struct net_buf`

duynguyenxa committed:
MAINTAINERS: Update Renesas files to cover mdio folder change

The mdio driver is moved into Ethernet driver folder
this change maintainer file to cover the change

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa commented on drivers/ethernet/eth_renesas_ra_rmac.c:
```diff
@@ -0,0 +1,754 @@
+		}
+
+		while(renesas_ra_eth_rx(dev)) {
+			k_yield();
```
Not really necessary, I just refer to some other implement, anyway I've removed it

duynguyenxa commented on boards/renesas/ek_ra8m2/ek_ra8m2_r7ka8m2jflcac_cm85.dts:
```diff
@@ -183,3 +184,35 @@
 
 mikrobus_spi: &spi1 {};
+
+&eswm {
```
All related boards update are added

duynguyenxa commented on drivers/ethernet/Kconfig.renesas_ra_rmac:
```diff
@@ -0,0 +1,48 @@
+	depends on DT_HAS_RENESAS_RA_ETHERNET_RMAC_ENABLED
+	select USE_RA_FSP_RMAC
+	select USE_RA_FSP_LAYER3_SWITCH
+	select MDIO
```
I've removed it

duynguyenxa commented on drivers/ethernet/eth_renesas_ra_rmac.c:
```diff
@@ -0,0 +1,753 @@
+	.en_irq = &renesas_ra_eswm_init_irq,
+};
+DEVICE_DT_INST_DEFINE(0, renesas_ra_eswm_init, NULL, &eswm_data, &eswm_config, POST_KERNEL,
+		      CONFIG_ESWM_RENESAS_RA_INIT_PRIORITY, NULL);
```
No, for now it just need to be initialize before the rmac and the mdio, I changed it to use  `CONFIG_ETH_INIT_PRIORITY` and remove `CONFIG_ESWM_RENESAS_RA_INIT_PRIORITY`

maass-hamburg commented on drivers/ethernet/mdio/Kconfig.renesas_ra:
```diff
@@ -9,3 +9,29 @@ config MDIO_RENESAS_RA
+	default 0
+	help
+	  Time for capturing MDI after rising edge of MDC in nanoseconds.
+	  This value depends on PHY chip AC Characteristics. Use 0 to disable.
```
if these are phy specific, i think they would make more sense as dt props of the mdio controller.

maass-hamburg commented on drivers/ethernet/phy/phy_maxlinear_gpy111.c:
```diff
@@ -0,0 +1,562 @@
+	uint32_t phy_id;
+	int ret = 0;
+
+	data->state.is_up = false;
```
line can be removed, false/0 already is the initial value

duynguyenxa commented on drivers/ethernet/eth_renesas_ra_rmac.c:
```diff
@@ -0,0 +1,753 @@
+				.array_length = ETH_TX_QUEUE_LEN(n),                               \
+				.p_descriptor_array = eth##n##_tx_desc_array##idx,                 \
+				.p_ts_descriptor_array = NULL,                                     \
+				.ports = 1 << DT_INST_PROP(n, channel),                            \
```
It's changed to BIT()

maass-hamburg commented on drivers/ethernet/eth_renesas_ra_rmac.c:
```diff
@@ -0,0 +1,754 @@
+
+#define ETH_USE_INTERNAL_PHY_CLK(id)                                                               \
+	COND_CODE_1(UTIL_AND(DT_NODE_HAS_COMPAT(id, renesas_ra_ethernet_rmac),                     \
+		 DT_ENUM_HAS_VALUE(id, phy_clock_type, internal)), (1), (0))
```
```suggestion
	UTIL_AND(DT_NODE_HAS_COMPAT(id, renesas_ra_ethernet_rmac),                     \
		 DT_ENUM_HAS_VALUE(id, phy_clock_type, internal)
```
no need for the COND_CODE_1 here

duynguyenxa commented on drivers/ethernet/mdio/mdio_renesas_ra_rmac.c:
```diff
@@ -0,0 +1,187 @@
+	BUILD_ASSERT(DT_INST_CHILD_NUM(n) <= BSP_FEATURE_ETHER_NUM_CHANNELS);                      \
+	PINCTRL_DT_INST_DEFINE(n);                                                                 \
+                                                                                                   \
+	static struct mdio_renesas_ra_data renesas_ra_mdio##n##_data = {0};                        \
```
it's removed
